TypeError: no unmapping handler for 0 of type <class 'int'>

Bumping `jax` leads to the following tests failures:
```
=========================== short test summary info ============================
FAILED tests/test_primitive.py::test_vprim - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_vmap_primal_unbatched_cond[lax-True] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_vmap_primal_batched_cond[lax-True] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_vmap_primal_unbatched_cond[bounded-True] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_vmap_primal_unbatched_cond[checkpointed-True] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_vmap_primal_batched_cond[checkpointed-True] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_speed_buffer_while[False] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_vmap_primal_batched_cond[bounded-True] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_speed_buffer_while[True] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_nested_loops[False] - TypeError: no unmapping handler for 0 of type <class 'int'>
FAILED tests/test_while_loop.py::test_nested_loops[True] - TypeError: no unmapping handler for 0 of type <class 'int'>
================== 11 failed, 448 passed, 1 skipped in 53.07s ==================
```

--- a/equinox/internal/_primitive.py
+++ b/equinox/internal/_primitive.py
@@ -344,6 +344,16 @@ def _vprim_impl(*inputs, prim, __axis_size, __axis_name, __batch_axes, params):
     return impl(*inputs)
 
 
+if jax.__version_info__ >= (0, 5, 1):
+
+    def _unmapped_aval(axis_size, axis_name, axis, aval):
+        del axis_name
+        return jax.core.unmapped_aval(axis_size, axis, aval)  # pyright: ignore[reportCallIssue]
+else:
+    # signature (axis_size, axis_name, axis, aval)
+    _unmapped_aval = jax.core.unmapped_aval  # pyright: ignore[reportAssignmentType]
+
+
 def _vprim_abstract_eval(*inputs, prim, __axis_size, __axis_name, __batch_axes, params):
     assert len(inputs) == len(__batch_axes)
     inputs = [
@@ -351,7 +361,7 @@ def _vprim_abstract_eval(*inputs, prim, __axis_size, __axis_name, __batch_axes, 
     ]
     abstract_eval = _vprim_abstract_eval_registry[prim]
     outs = abstract_eval(*inputs, **params)
-    outs = [jax.core.unmapped_aval(__axis_size, __axis_name, 0, x) for x in outs]
+    outs = [_unmapped_aval(__axis_size, __axis_name, 0, x) for x in outs]
     return outs
 
 

--- a/equinox/nn/_activations.py
+++ b/equinox/nn/_activations.py
@@ -5,6 +5,7 @@ import jax.numpy as jnp
 from jaxtyping import Array
 
 from .._module import Module
+from ._misc import named_scope
 
 
 class PReLU(Module, strict=True):
@@ -30,7 +31,7 @@ class PReLU(Module, strict=True):
 
         self.negative_slope = jnp.asarray(init_alpha)
 
-    @jax.named_scope("eqx.nn.PReLU")
+    @named_scope("eqx.nn.PReLU")
     def __call__(self, x: Array) -> Array:
         r"""**Arguments:**
 

--- a/equinox/nn/_attention.py
+++ b/equinox/nn/_attention.py
@@ -14,6 +14,7 @@ from .._misc import default_floating_dtype
 from .._module import field, Module
 from ._dropout import Dropout
 from ._linear import Linear
+from ._misc import named_scope
 
 
 def dot_product_attention_weights(
@@ -234,7 +235,7 @@ class MultiheadAttention(Module, strict=True):
         self.use_value_bias = use_value_bias
         self.use_output_bias = use_output_bias
 
-    @jax.named_scope("eqx.nn.MultiheadAttention")
+    @named_scope("eqx.nn.MultiheadAttention")
     def __call__(
         self,
         query: Float[Array, "q_seq q_size"],

--- a/equinox/nn/_batch_norm.py
+++ b/equinox/nn/_batch_norm.py
@@ -8,6 +8,7 @@ from jaxtyping import Array, Bool, Float, PRNGKeyArray
 
 from .._misc import default_floating_dtype
 from .._module import field
+from ._misc import named_scope
 from ._sequential import StatefulLayer
 from ._stateful import State, StateIndex
 
@@ -106,7 +107,7 @@ class BatchNorm(StatefulLayer, strict=True):
         self.channelwise_affine = channelwise_affine
         self.momentum = momentum
 
-    @jax.named_scope("eqx.nn.BatchNorm")
+    @named_scope("eqx.nn.BatchNorm")
     def __call__(
         self,
         x: Array,

--- a/equinox/nn/_conv.py
+++ b/equinox/nn/_conv.py
@@ -3,7 +3,6 @@ import math
 from collections.abc import Callable, Sequence
 from typing import Optional, TypeVar, Union
 
-import jax
 import jax.lax as lax
 import jax.numpy as jnp
 import jax.random as jrandom
@@ -12,7 +11,7 @@ from jaxtyping import Array, PRNGKeyArray
 
 from .._misc import default_floating_dtype
 from .._module import field, Module
-from ._misc import all_sequences, default_init
+from ._misc import all_sequences, default_init, named_scope
 
 
 _T = TypeVar("_T")
@@ -208,7 +207,7 @@ class Conv(Module, strict=True):
         x = jnp.pad(x, [(0, 0)] + padding, mode)
         return x
 
-    @jax.named_scope("eqx.nn.Conv")
+    @named_scope("eqx.nn.Conv")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -580,7 +579,7 @@ class ConvTranspose(Module, strict=True):
         padding_t = tuple((p[0].item(), p[1].item()) for p in padding_t % stride)
         return x, padding_t
 
-    @jax.named_scope("eqx.nn.ConvTranspose")
+    @named_scope("eqx.nn.ConvTranspose")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 

--- a/equinox/nn/_dropout.py
+++ b/equinox/nn/_dropout.py
@@ -1,13 +1,13 @@
 import warnings
 from typing import Optional
 
-import jax
 import jax.lax as lax
 import jax.numpy as jnp
 import jax.random as jrandom
 from jaxtyping import Array, PRNGKeyArray
 
 from .._module import Module
+from ._misc import named_scope
 
 
 class Dropout(Module, strict=True):
@@ -54,7 +54,7 @@ class Dropout(Module, strict=True):
     def deterministic(self):
         return self.inference
 
-    @jax.named_scope("eqx.nn.Dropout")
+    @named_scope("eqx.nn.Dropout")
     def __call__(
         self,
         x: Array,

--- a/equinox/nn/_embedding.py
+++ b/equinox/nn/_embedding.py
@@ -10,6 +10,7 @@ from .._caches import cache_clears
 from .._filters import is_array_like
 from .._misc import default_floating_dtype
 from .._module import field, Module
+from ._misc import named_scope
 
 
 internal_rope_embedding_cache: dict[tuple[int, Any], tuple[Array, Array]] = {}
@@ -81,7 +82,7 @@ class Embedding(Module, strict=True):
         self.num_embeddings = num_embeddings
         self.embedding_size = embedding_size
 
-    @jax.named_scope("eqx.nn.Embedding")
+    @named_scope("eqx.nn.Embedding")
     def __call__(
         self, x: Int[ArrayLike, ""], *, key: Optional[PRNGKeyArray] = None
     ) -> Array:
@@ -191,7 +192,7 @@ class RotaryPositionalEmbedding(Module, strict=True):
         # we assign the type at the very end to minimize the loss of precision
         return jnp.cos(freqs_outer).astype(dtype), jnp.sin(freqs_outer).astype(dtype)
 
-    @jax.named_scope("eqx.nn.RotaryPositionalEmbedding")
+    @named_scope("eqx.nn.RotaryPositionalEmbedding")
     def __call__(
         self,
         x: Float[Array, "seq_length embedding_size"],

--- a/equinox/nn/_linear.py
+++ b/equinox/nn/_linear.py
@@ -1,14 +1,13 @@
 import math
 from typing import Any, Literal, Optional, TypeVar, Union
 
-import jax
 import jax.numpy as jnp
 import jax.random as jrandom
 from jaxtyping import Array, PRNGKeyArray
 
 from .._misc import default_floating_dtype
 from .._module import field, Module
-from ._misc import default_init
+from ._misc import default_init, named_scope
 
 
 class Linear(Module, strict=True):
@@ -65,7 +64,7 @@ class Linear(Module, strict=True):
         self.out_features = out_features
         self.use_bias = use_bias
 
-    @jax.named_scope("eqx.nn.Linear")
+    @named_scope("eqx.nn.Linear")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -115,7 +114,7 @@ class Identity(Module, strict=True):
     def __init__(self, *args: Any, **kwargs: Any):
         """Consumes arbitrary `*args` and `**kwargs` but ignores them."""
 
-    @jax.named_scope("eqx.nn.Identity")
+    @named_scope("eqx.nn.Identity")
     def __call__(self, x: _T, *, key: Optional[PRNGKeyArray] = None) -> _T:
         """**Arguments:**
 

--- a/equinox/nn/_misc.py
+++ b/equinox/nn/_misc.py
@@ -1,5 +1,5 @@
 import typing_extensions as te
-from collections.abc import Sequence
+from collections.abc import Callable, Sequence
 from typing import Any, TYPE_CHECKING, TypeVar, Union
 
 import jax
@@ -18,10 +18,16 @@ if TYPE_CHECKING:
         x: Union[Sequence[Any], Sequence[_T]],
     ) -> "te.TypeIs[Sequence[_T]]": ...
 
+    _S = TypeVar("_S")
+
+    def named_scope(name: str) -> Callable[[_S], _S]: ...
+
 else:
     # beartype doesn't like StrictTypeGuard
     def all_sequences(x: Union[Sequence[Any], Sequence[_T]]) -> bool:
         return all(isinstance(xi, Sequence) for xi in x)
+
+    named_scope = jax.named_scope
 
 
 def default_init(

--- a/equinox/nn/_mlp.py
+++ b/equinox/nn/_mlp.py
@@ -5,7 +5,6 @@ from typing import (
     Union,
 )
 
-import jax
 import jax.nn as jnn
 import jax.random as jrandom
 import jax.tree_util as jtu
@@ -17,6 +16,7 @@ from .._misc import default_floating_dtype
 from .._module import field, Module
 from .._vmap_pmap import filter_vmap
 from ._linear import Linear
+from ._misc import named_scope
 
 
 _identity = doc_repr(lambda x: x, "lambda x: x")
@@ -125,7 +125,7 @@ class MLP(Module, strict=True):
         self.use_bias = use_bias
         self.use_final_bias = use_final_bias
 
-    @jax.named_scope("eqx.nn.MLP")
+    @named_scope("eqx.nn.MLP")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 

--- a/equinox/nn/_normalisation.py
+++ b/equinox/nn/_normalisation.py
@@ -10,6 +10,7 @@ from jaxtyping import Array, Float, PRNGKeyArray
 from .._custom_types import sentinel
 from .._misc import default_floating_dtype, left_broadcast_to
 from .._module import field, Module
+from ._misc import named_scope
 from ._stateful import State
 
 
@@ -105,7 +106,7 @@ class LayerNorm(Module, strict=True):
         self, x: Array, state: State, *, key: Optional[PRNGKeyArray] = None
     ) -> tuple[Array, State]: ...
 
-    @jax.named_scope("eqx.nn.LayerNorm")
+    @named_scope("eqx.nn.LayerNorm")
     def __call__(
         self,
         x: Float[Array, "*shape"],
@@ -238,7 +239,7 @@ class GroupNorm(Module, strict=True):
         self, x: Array, state: State, *, key: Optional[PRNGKeyArray] = None
     ) -> tuple[Array, State]: ...
 
-    @jax.named_scope("eqx.nn.GroupNorm")
+    @named_scope("eqx.nn.GroupNorm")
     def __call__(
         self, x: Array, state: State = sentinel, *, key: Optional[PRNGKeyArray] = None
     ) -> Union[Array, tuple[Array, State]]:
@@ -357,7 +358,7 @@ class RMSNorm(Module, strict=True):
         self, x: Array, state: State, *, key: Optional[PRNGKeyArray] = None
     ) -> tuple[Array, State]: ...
 
-    @jax.named_scope("eqx.nn.RMSNorm")
+    @named_scope("eqx.nn.RMSNorm")
     def __call__(
         self,
         x: Float[Array, "*shape"],

--- a/equinox/nn/_pool.py
+++ b/equinox/nn/_pool.py
@@ -9,7 +9,7 @@ import jax.random
 from jaxtyping import Array, PRNGKeyArray
 
 from .._module import field, Module
-from ._misc import all_sequences
+from ._misc import all_sequences, named_scope
 
 
 class Pool(Module, strict=True):
@@ -114,7 +114,7 @@ class Pool(Module, strict=True):
                     f"{kernel_size}."
                 )
 
-    @jax.named_scope("eqx.nn.Pool")
+    @named_scope("eqx.nn.Pool")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -186,7 +186,7 @@ class AvgPool1d(Pool):
             use_ceil=use_ceil,
         )
 
-    @jax.named_scope("eqx.nn.AvgPool1d")
+    @named_scope("eqx.nn.AvgPool1d")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -234,7 +234,7 @@ class MaxPool1d(Pool):
         )
 
     # Redefined to get them in the right order in docs
-    @jax.named_scope("eqx.nn.MaxPool1d")
+    @named_scope("eqx.nn.MaxPool1d")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -281,7 +281,7 @@ class AvgPool2d(Pool):
             use_ceil=use_ceil,
         )
 
-    @jax.named_scope("eqx.nn.AvgPool2d")
+    @named_scope("eqx.nn.AvgPool2d")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -329,7 +329,7 @@ class MaxPool2d(Pool):
         )
 
     # Redefined to get them in the right order in docs
-    @jax.named_scope("eqx.nn.MaxPool2d")
+    @named_scope("eqx.nn.MaxPool2d")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -376,7 +376,7 @@ class AvgPool3d(Pool):
             use_ceil=use_ceil,
         )
 
-    @jax.named_scope("eqx.nn.AvgPool3d")
+    @named_scope("eqx.nn.AvgPool3d")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -424,7 +424,7 @@ class MaxPool3d(Pool):
             use_ceil=use_ceil,
         )
 
-    @jax.named_scope("eqx.nn.MaxPool3d")
+    @named_scope("eqx.nn.MaxPool3d")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 
@@ -502,7 +502,7 @@ class AdaptivePool(Module, strict=True):
                 f"{num_spatial_dims} containing ints."
             )
 
-    @jax.named_scope("eqx.nn.AdaptivePool")
+    @named_scope("eqx.nn.AdaptivePool")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 

--- a/equinox/nn/_rnn.py
+++ b/equinox/nn/_rnn.py
@@ -1,7 +1,6 @@
 import math
 from typing import Optional
 
-import jax
 import jax.nn as jnn
 import jax.numpy as jnp
 import jax.random as jrandom
@@ -9,7 +8,7 @@ from jaxtyping import Array, PRNGKeyArray
 
 from .._misc import default_floating_dtype
 from .._module import field, Module
-from ._misc import default_init
+from ._misc import default_init, named_scope
 
 
 class GRUCell(Module, strict=True):
@@ -82,7 +81,7 @@ class GRUCell(Module, strict=True):
         self.hidden_size = hidden_size
         self.use_bias = use_bias
 
-    @jax.named_scope("eqx.nn.GRUCell")
+    @named_scope("eqx.nn.GRUCell")
     def __call__(
         self, input: Array, hidden: Array, *, key: Optional[PRNGKeyArray] = None
     ):
@@ -178,7 +177,7 @@ class LSTMCell(Module, strict=True):
         self.hidden_size = hidden_size
         self.use_bias = use_bias
 
-    @jax.named_scope("eqx.nn.LSTMCell")
+    @named_scope("eqx.nn.LSTMCell")
     def __call__(self, input, hidden, *, key=None):
         """**Arguments:**
 

--- a/equinox/nn/_sequential.py
+++ b/equinox/nn/_sequential.py
@@ -1,13 +1,13 @@
 from collections.abc import Callable, Sequence
 from typing import Any, Optional, overload, Union
 
-import jax
 import jax.random as jr
 from jaxtyping import Array, PRNGKeyArray
 
 from .._better_abstract import AbstractClassVar
 from .._custom_types import sentinel
 from .._module import Module, StrictConfig
+from ._misc import named_scope
 from ._stateful import State
 
 
@@ -71,7 +71,7 @@ class Sequential(StatefulLayer, strict=StrictConfig(allow_method_override=True))
         self, x: Array, state: State, *, key: Optional[PRNGKeyArray] = None
     ) -> tuple[Array, State]: ...
 
-    @jax.named_scope("eqx.nn.Sequential")
+    @named_scope("eqx.nn.Sequential")
     def __call__(
         self,
         x: Array,

--- a/equinox/nn/_spectral_norm.py
+++ b/equinox/nn/_spectral_norm.py
@@ -1,6 +1,5 @@
 from typing import Generic, Optional, TypeVar
 
-import jax
 import jax.lax as lax
 import jax.numpy as jnp
 import jax.random as jr
@@ -8,6 +7,7 @@ from jaxtyping import Array, Float, PRNGKeyArray
 
 from .._module import field
 from .._tree import tree_at
+from ._misc import named_scope
 from ._sequential import StatefulLayer
 from ._stateful import State, StateIndex
 
@@ -110,7 +110,7 @@ class SpectralNorm(StatefulLayer, Generic[_Layer], strict=True):
             u0, v0 = _power_iteration(weight, u0, v0, eps)
         self.uv_index = StateIndex((u0, v0))
 
-    @jax.named_scope("eqx.nn.SpectralNorm")
+    @named_scope("eqx.nn.SpectralNorm")
     def __call__(
         self,
         x: Array,

--- a/equinox/nn/_weight_norm.py
+++ b/equinox/nn/_weight_norm.py
@@ -8,6 +8,7 @@ from jaxtyping import Array, PRNGKeyArray, Scalar
 
 from .._module import field, Module
 from .._tree import tree_at
+from ._misc import named_scope
 
 
 _Layer = TypeVar("_Layer")
@@ -94,7 +95,7 @@ class WeightNorm(Module, Generic[_Layer], strict=True):
         )
         self.g = self._norm(getattr(layer, weight_name))
 
-    @jax.named_scope("eqx.nn.WeightNorm")
+    @named_scope("eqx.nn.WeightNorm")
     def __call__(self, x: Array, *, key: Optional[PRNGKeyArray] = None) -> Array:
         """**Arguments:**
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,6 @@
 [project]
 name = "equinox"
-version = "0.11.11"
+version = "0.11.12"
 description = "Elegant easy-to-use neural networks in JAX."
 readme = "README.md"
 requires-python =">=3.10"

